rust/core-lib/src/recorder.rs

```diff
@@ -0,0 +1,435 @@
+// Copyright 2018 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use xi_trace::trace_block;
+
+use std::mem;
+use std::collections::HashMap;
+
+use edit_types::{BufferEvent, EventDomain};
+
+/// A container that manages and holds all recordings for the current editing session
+pub(crate) struct Recorder {
+    active_recording: Option<String>,
+    recording_buffer: Vec<EventDomain>,
+    recordings: HashMap<String, Recording>,
+}
+
+impl Recorder {
+    pub(crate) fn new() -> Recorder {
+        Recorder {
+            active_recording: None,
+            recording_buffer: Vec::new(),
+            recordings: HashMap::new(),
+        }
+    }
+
+    pub(crate) fn is_recording(&self) -> bool {
+        self.active_recording.is_some()
+    }
+
+    /// Starts or stops the specified recording.
+    ///
+    ///
+    /// There are three outcome behaviors:
+    /// - If the current recording name is specified, the active recording is saved
+    /// - If no recording name is specified, the currently active recording is saved
+    /// - If a recording name other than the active recording is specified,
+    /// the current recording will be thrown out and will be switched to the new name
+    ///
+    /// In addition to the above:
+    /// - If the recording was saved, there is no active recording
+    /// - If the recording was switched, there will be a new active recording
+    pub(crate) fn toggle_recording(&mut self, recording_name: Option<String>) {
+        let is_recording = self.is_recording();
+        let last_recording = self.active_recording.take();
+
+        match (is_recording, &last_recording, &recording_name) {
+            (true, Some(last_recording), None) => self.save_recording_buffer(last_recording.clone()),
```

Comment:
Unfortunately, yes since we can't bind a ref and a move in the same pattern (https://doc.rust-lang.org/error-index.html#E0009).

We have the borrow for `recording_name` since we also use it for the `mem::replace` at the bottom of the function.